### DISCUSSION TITLE
Cannot install any version of Ruby - M1 Max / Xcode 13.2.1 / Monterey 12.2.1

### DISCUSSION THREAD

**Discussion by iotashan** [rbenv/ruby-build#1947] [Build failures] [answered]
At some point in time since I got my new M1-Max MBP in February, installing new versions of Ruby has broken. I can't seem to find any info online about this "cannot run C compiled programs" that's not referencing issues solved 5+ years ago.

```zsh
➜  ~ rbenv install 2.7.5               
Downloading openssl-1.1.1l.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/0b7a3e5e59c34827fe0c3a74b7ec8baef302b98fa80088d7f9153aa16fa76bd1
Installing openssl-1.1.1l...
Installed openssl-1.1.1l to /Users/shan/.rbenv/versions/2.7.5

Downloading ruby-2.7.5.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.5.tar.bz2
Installing ruby-2.7.5...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.2.1 using ruby-build 20220218)

Inspect or clean up the working tree at /var/folders/6d/vhw47jn57kj08mhpwcn2ymr80000gn/T/ruby-build.20220303160033.86276.m8tbpl
Results logged to /var/folders/6d/vhw47jn57kj08mhpwcn2ymr80000gn/T/ruby-build.20220303160033.86276.log

Last 10 log lines:
checking host system type... aarch64-apple-darwin21.3.0
checking target system type... aarch64-apple-darwin21.3.0
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/var/folders/6d/vhw47jn57kj08mhpwcn2ymr80000gn/T/ruby-build.20220303160033.86276.m8tbpl/ruby-2.7.5':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
make: *** No targets specified and no makefile found.  Stop.
```

I get the same overall error if I try to install 3.0.3. The worst thing about this is that I had previously installed 2.7.5 and everything was working well, but then when I tried to re-install 2.7.5 it broke the existing install.

**Comment by mislav** [comment]
No help here, sorry. For what it's worth, I've tried compiling 2.7.5 on a fresh M1 machine and it works out of the box without problems. I do not have access to a M1 Max machine.

**Comment by hsbt** [comment]
I could install them with the same environment, which is M1 Max, macOS 12.2.1. 

Maybe you mixed the x86 and aarch64 architecture with Rosetta2.

**Reply by iotashan** [reply]
I've been so careful to avoid doing any x86 stuff. All my tools are M1 native now. I'll go back and re-check everything.

Might have to rip it all out and reinstall it again.

**Comment by iotashan** [comment, marked as answer]
I ended up doing the following:

1. `brew remove rbenv ruby-build`
2. `rm -rf ~/.rbenv`
3. `rm -rf ~/Library/Caches/Homebrew`
4. `brew update && brew upgrade`

After this, I was able to install Ruby again. No idea what the problem was.

**Reply by phorsuedzie** [reply]
> No idea what the problem was.

If the config.log (above all the debug information) shows "killed" (with signal: 9), then it's the [system integrity protection](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection) that hurts the install: macOS doesn't allow the recently compiled program to be executed, but kills it.

I had to re-install the CommandlineTools to get it running eventually:

```
sudo rm -rf /Library/Developers/CommandLineTools
xcode-select --install
```

Note: to compile (e.g. `rbenv install 3.2.2`) with system integrity protection disabled  "succeeded". But `ruby -v` was still `Killed: 9` afterwards.

**Comment by kanishkdailypay** [comment]
Try ! Rebooting ! It really helps and can solve the issue !
